### PR TITLE
Extended flux

### PR DESCRIPTION
**Pull request by henrikef**
Hi! 

I made some changes to both astromodels and threeML to allow to assign the XYLike plugin to an extended source. I want to use this for example in case we have spectral measurements of an extended source from IACTs or Fermi that we want to use via the XYLike plugin. Obviously,  this would rely on the user to  pick the 'right' spatial shape when fitting with any plugins with spatial resolution. The change to astromodels is just an additional function in the extended_source class, which returns the total flux at a given energy.

### CODE REVIEW

**Comment by giacomov** [comment]
Thanks, it is a very nice addition.

However, I believe there is a better strategy. Indeed, It looks to me that the method you added (get_total_flux) makes sense only for sources that have the same spectrum everywhere. Hence, it is not general enough to be in extended_source.py because it would fail or give not meaningful results for other kinds of sources.

Moreover, the new method is largely a copy paste from the __call__ method. Copy/paste (i.e., duplication of code) should be avoided at all costs.

I have a different proposal to achieve the same result with no modification to astromodels.

The current calling sequence of the `__init__` method of `XYLike` is:

```python
def __init__(self, name, x, y, yerr=None, poisson_data=False, quiet=False, source_name=None):
```

You can add a new optional keyword there like this:

```python
def __init__(self, name, x, y, yerr=None, poisson_data=False, quiet=False, source_name=None, ext_source_flux_position=None):
```

The user will need to specify that as a (ra, dec) tuple when using XYLike with an extended source. Then, you will be able to get the flux from the extended sources as:

```python
assert self._ext_source_flux_position is not None, "When using XYLike for an extended source you need to specify a position where to evaluate the source at"
lon, lat = self._ext_source_flux_position
expectation = self._likelihood_model.extended_sources[self._source_name].(lon, lat, self._x)
```

This won't require any change to astromodels and will work for any kind of extended source, even those with position-dependent spectrum.

**Comment by henrikef** [comment]
Hi @giacomov ! I can add text and fix the code duplication issue. I was trying to leave the existing functions alone as much as possible, but I guess duplicating code makes it harder to maintain.

As for your suggestions to add the ra, dec keyword to XYLike: I agree that it would be more elegant, but I don't think  it's what I want. 

`self._likelihood_model.extended_sources[self._source_name].(lon, lat, self._x) ` gives me the flux coming from a particular direction on the sky (let's call it f =  dN/(dE dt dA dOmega) ). If I look at e.g.  VERITAS  measurements of an extended source, they claim to measure what I called the 'total' flux, F = dN/(dE dt dA) = integral f dOmega . (Obviously this assumes that they chose their integration region large enough, otherwise it will be the integral over some part of the support of f).

In any case, for a point source, the 'spatial' part of f is a delta function, which we tend to ignore. But for an extended source, we cannot really do that. For example, the units shouldn't be correct if I replace the  'total' flux by the flux at the center of the source.

Note that even for sources with non-constant  energy spectrum, the total flux F is well defined. For sources with separable spatial and spectral components, accessing the spectral component via my  `get_total_flux` function should be enough to give us F as long as the spatial component is normalized properly, which as far as I  know all the 2D functions in astromodels are. The only such functions  I'm aware of currently in astromodels are the continuus_injection_diffusion ones. I'm not sure about the normalization there but as far as I am aware, the quoted spectral parameters e.g. for Geminga are just the spectral parts of the model.

The only sources for which my code would fail are ones which do not have separable spectral and spatial components. For those sources, F is still well defined but would have to be calculated by explicitly integrating over the spatial part. If you could give me an example of any such source/model, I can try make it work for those.

Btw, the same goes for the 'plotting' branch: Even if you/we add an option to plot the spectrum at a particular direction, I'd still want the option to plot  the 'total' spectrum (i.e. integrating over the spatial part) to be able  to compare to other instruments' results.

File: astromodels/functions/functions_2D.py [hunk 1]
```diff
@@ -95,7 +95,23 @@ def get_boundaries(self):
...
+        
+    def get_total_spatial_integral(self, z=None):  
+        """
+        Returns the total integral (for 2D functions) or the integral over the spatial components (for 3D functions).
+        needs to be implemented in subclasses.
+
+        :return: an array of values of the integral (same dimension as z).
+        """
+
+        dL= l_max-l_min if l_max > l_min else 360 + l_max - l_max
```
Comment (giacomov) [inline astromodels/functions/functions_2D.py]: Looks like there are a bunch of undefined variables in this method (`l_max`, `l_min`, `sigma_b`...). Did you test it?

File: astromodels/functions/functions_2D.py [hunk 2]
```diff
@@ -629,6 +683,17 @@ def get_boundaries(self):
...
 
+    def get_total_spatial_integral(self, z=None):  
+        """
+        Returns the total integral (for 2D functions) or the integral over the spatial components (for 3D functions).
+        needs to be implemented in subclasses.
+
+        :return: an array of values of the integral (same dimension as z).
+        """
+
+        return K
```
Comment (giacomov) [inline astromodels/functions/functions_2D.py]: where is `K` defined in the local scope?

**Pull request merged**